plug.vim

```diff
@@ -968,6 +973,377 @@ while 1 " Without TCO, Vim stack is bound to explode
+    with self.lock:
+      self._write(*args, **kwargs)
+
+  def _write(self, action, name, lines):
+    """ Main logging function, updates based on threads calling. """
+    if action == Action.ERROR:
+      self.bar += 'x'
+      vim.command("call add(s:update.errors, '{}')".format(name))
+    elif action == Action.DONE:
+      self.bar += '='
+
+    first, rest = lines[0], lines[1:]
+    msg = ['{} {}: {}'.format(action, name, first)]
+    padded_rest = ['    ' + line for line in rest]
+    msg.extend(padded_rest)
+
+    curbuf = vim.current.buffer
+    lnum = self.where(name)
+    if lnum != -1: # Found matching line num
+      del curbuf[lnum]
+      if lnum > self.maxy and action in {Action.INSTALL, Action.UPDATE}:
+        lnum = 3
+    else:
+      lnum = 3
+    curbuf.append(msg, lnum)
+
+    self.header()
+
+class Command(object):
+  def __init__(self, cmd, cmd_dir=None, timeout=60, ntries=3, cb=None, clean=None):
+    self.cmd = cmd
+    self.cmd_dir = cmd_dir
+    self.timeout = timeout
+    self.ntries = ntries
+
+    self.clean = clean
+    if cb:
+      self.callback = cb
+    else:
+      def dummy(msg): pass
+      self.callback = dummy
+
+  def attempt_cmd(self):
+    """ Looping call to attempt cmd ntries with a timeout. """
+    attempt = 0
+    finished = False
+    limit = self.timeout
+
+    while not finished:
+      try:
+        attempt += 1
+        result = self.timeout_cmd()
+        finished = True
+      except CmdTimedOut:
+        if attempt != self.ntries:
+          for count in range(3, 0, -1):
+            msg = 'Timeout. Will retry in {} second{} ...'.format(
+                count, 's' if count != 1 else '')
+            self.callback([msg])
+            time.sleep(1)
+          self.timeout += limit
+          self.callback(['Retrying ...'])
+        else:
+          raise
+
+    return result
+
+  def timeout_cmd(self):
+    """ Execute a cmd and returns output. Output is list of lines. BLOCKING.
+    Raises CmdFailed   -> return code for Popen isn't 0
+    Raises CmdTimedOut -> command exceeded timeout without new output
+    """
+    pcmd = None
+    first_line = True
+    try:
+      tfile = tempfile.NamedTemporaryFile()
+      pcmd = subprocess.Popen(self.cmd, cwd=self.cmd_dir, stdout=tfile,
+          stderr=subprocess.STDOUT, shell=True)
+      while pcmd.poll() == None:
+        # Yield this thread
+        time.sleep(0)
+
+        if first_line or random.random() < G_LOG_PROB:
+          first_line = False
+          line = nonblock_read(tfile.name)
+          if line and self.callback:
+            self.callback([line])
+
+        time_diff = time.time() - os.path.getmtime(tfile.name)
+        if time_diff > self.timeout:
+          raise CmdTimedOut(['Timeout!'])
+
+      tfile.seek(0)
+      result = [line.rstrip() for line in tfile]
+
+      if pcmd.returncode != 0:
+        msg = ['CmdFailed:', self.cmd]
+        msg.extend(result)
+        raise CmdFailed(msg)
+    except:
```

Comment:
A bare `except` is probably not what you want. Change this to:

```python
except Exception:
```

>A bare `except:` clause will catch SystemExit and KeyboardInterrupt exceptions, making it harder to interrupt a program with Control-C, and can disguise other problems. If you want to catch all exceptions that signal program errors, use `except Exception:`

See https://www.python.org/dev/peps/pep-0008/ (under "Programming Recommendations") for more information about this.